Applications/Spire/source/charting/trend_line_model.cpp

```diff
@@ -0,0 +1,193 @@
+  for(auto& line : m_trend_lines) {
+    auto dist_squared = std::numeric_limits<Quantity>::infinity();
+    auto point_distance_squared = std::numeric_limits<Quantity>::infinity();
+    auto line_slope = slope(line.m_trend_line);
+    auto line_b = y_intercept(
+      static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_x),
+      static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_y),
+      line_slope);
+    point_distance_squared = closest_point_distance_squared(point_x,
+      point_y, line.m_trend_line);
+    if(std::isinf(static_cast<double>(line_slope))) {
+      if(is_within_interval(point_y,
+          static_cast<Quantity>(std::get<0>(
+          line.m_trend_line.m_points).m_y),
+          static_cast<Quantity>(std::get<1>(
+          line.m_trend_line.m_points).m_y))) {
+        dist_squared = distance_squared(point_x, point_y,
+          static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_x),
+          point_y);
+      }
+    } else if(line_slope == Quantity(0)) {
+      if(is_within_interval(point_x,
+          static_cast<Quantity>(std::get<0>(
+          line.m_trend_line.m_points).m_x),
+          static_cast<Quantity>(std::get<1>(
+          line.m_trend_line.m_points).m_x))) {
+        dist_squared = distance_squared(point_x, point_y, point_x,
+          static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_y));
+      }
+    } else {
+      auto line_point_x =
+        (point_x + (line_slope * point_y) - (line_slope * line_b)) /
+        ((line_slope * line_slope) + 1);
+      if(is_within_interval(line_point_x,
+          static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_x),
+          static_cast<Quantity>(std::get<1>(line.m_trend_line.m_points).m_x))) {
+        dist_squared = distance_squared(point_x, point_y,
+          line_point_x, calculate_y(line_slope, line_point_x, line_b));
+      }
+    }
+    auto min_distance_squared = std::min(dist_squared, point_distance_squared);
+    if(min_distance_squared <= threshold_squared &&
+        min_distance_squared < closest_distance) {
+      closest_id = line.m_id;
+      closest_distance = min_distance_squared;
+    }
+  }
+  return closest_id;
+}
+
+void TrendLineModel::remove(int id) {
+  auto iter = find_id(id);
+  if(iter != m_trend_lines.end()) {
+    m_trend_lines.erase(iter);
+  }
+}
+
+void TrendLineModel::set_selected(int id) {
+  auto iter = find_id(id);
+  if(iter != m_trend_lines.end()) {
+    iter->m_state = State::SELECTED;
+  }
+}
+
+void TrendLineModel::unset_selected(int id) {
+  auto iter = find_id(id);
+  if(iter != m_trend_lines.end()) {
+    iter->m_state = State::UNSELECTED;
+  }
+}
+
+void TrendLineModel::update(const TrendLine& line, int id) {
+  auto iter = find_id(id);
+  if(iter != m_trend_lines.end()) {
+    iter->m_trend_line = line;
+  }
+}
+
+connection TrendLineModel::connect_update_signal(
+    const UpdateSignal::slot_type& slot) const {
+  return m_update_signal.connect(slot);
```

Comment:
This signal is never being emitted. It should serve a purpose or be removed from the model.